### PR TITLE
Use const generics for hash_data_cache mmap

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

hash_data_cache mmap header and cell sizes are always constant. We can use const generics instead of member variable. 

#### Summary of Changes

Use const generic for `CacheHashDataFile`.

Fixes #

### CODE REVIEW

**Comment by alessandrod** [comment]
This PR does three things:

1) it adds explicit padding to `Header` so that **what follows** it is guaranteed to be aligned correctly

I touched on this on the previous review: this is technically not required now since `size_of::<Header>() == size_of::<usize>()` so the first byte _after_ `Header` is already aligned to the cell size. But if the `Header` layout changes, then the fist byte past it might become unaligned. 

The way the alignment is enforced is also problematic tho: `align: [u8; CELL_ALIGNMENT - std::mem::size_of::<usize>()]`. If another field is added, then the `_align` field needs to be modified too to take into account the size of the new field.

I'd rather just add a comment to `Header` saying: `NOTE: if you add or remove fields, you MUST ensure that size_of::<Header>() is equal or a multiple of CELL_SIZE`. And iirc we have a `debug_assert` somewhere to ensure this so tests will hopefully fail otherwise. 

2) It makes `CELL_SIZE` and `CELL_ALIGNMENT` constant

This seems fine to me.

3) It adds const generics to `CacheHashDataFile`

I understand why but I don't think this is actually needed. Are we going to use `CacheHashDataFile` to store different data with different layouts? Seems unlikely to me.

**Comment by alessandrod** [comment]
> The way the alignment is enforced is also problematic tho: align: [u8; CELL_ALIGNMENT - std::mem::size_of::<usize>()]. If another field is added, then the _align field needs to be modified too to take into account the size of the new field.

If you wanted to align in code, this is how I'd recommend doing it https://docs.rs/aya/latest/src/aya/util.rs.html#128-146

File: runtime/src/cache_hash_data.rs [hunk 1]
```diff
@@ -19,18 +19,22 @@ pub type EntryType = CalculateHashIntermediate;
 pub type SavedType = Vec<Vec<EntryType>>;
 pub type SavedTypeSlice = [Vec<EntryType>];
 
+const CELL_SIZE: usize = std::mem::size_of::<EntryType>();
+const CELL_ALIGNMENT: usize = std::mem::align_of::<EntryType>();
+
 #[repr(C)]
 pub struct Header {
     count: usize,
+    _align: [u8; CELL_ALIGNMENT - std::mem::size_of::<usize>()],
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply runtime/src/cache_hash_data.rs:28]: For usize = u64, it does nothing. If usize=u32, it will add 4 bytes?

**Comment by HaoranYi** [comment]
> This PR does three things:
> 
> 1. it adds explicit padding to `Header` so that **what follows** it is guaranteed to be aligned correctly
> 
> I touched on this on the previous review: this is technically not required now since `size_of::<Header>() == size_of::<usize>()` so the first byte _after_ `Header` is already aligned to the cell size. But if the `Header` layout changes, then the fist byte past it might become unaligned.
> 
> The way the alignment is enforced is also problematic tho: `align: [u8; CELL_ALIGNMENT - std::mem::size_of::<usize>()]`. If another field is added, then the `_align` field needs to be modified too to take into account the size of the new field.
> 
> I'd rather just add a comment to `Header` saying: `NOTE: if you add or remove fields, you MUST ensure that size_of::<Header>() is equal or a multiple of CELL_SIZE`. And iirc we have a `debug_assert` somewhere to ensure this so tests will hopefully fail otherwise.
> 
> 2. It makes `CELL_SIZE` and `CELL_ALIGNMENT` constant
> 
> This seems fine to me.
> 
> 3. It adds const generics to `CacheHashDataFile`
> 
> I understand why but I don't think this is actually needed. Are we going to use `CacheHashDataFile` to store different data with different layouts? Seems unlikely to me.

Thanks for the recap. 
My intention is to refactor `CacheHashDataFile` into a generic `CacheDataFile` for any EntryType.

File: runtime/src/cache_hash_data.rs [hunk 2]
```diff
@@ -62,51 +67,38 @@ impl CacheHashDataFile {
...
-        let remaining_elements = item_slice.len() / std::mem::size_of::<EntryType>();
+        let remaining_elements = item_slice.len() / CELL_SIZE;
         unsafe {
             let item = item_slice.as_ptr() as *const EntryType;
             std::slice::from_raw_parts(item, remaining_elements)
         }
     }
 
+    /// get '&mut [EntryType]' from cache file [ix..]
+    fn get_slice_mut(&self, ix: u64) -> &mut [EntryType] {
```
Comment (alessandrod) [inline runtime/src/cache_hash_data.rs:82]: I see you're still working on it and I'll do a proper review when you're done. I'll point out that this is undefined behaviour tho. You need to make `get_slice_mut` take `&mut self` or it becomes possible to have aliasing mut references which is UB.

**Pull request closed without merging**